pkg/config/engine/containerd/cgroup_driver.go

```diff
@@ -0,0 +1,53 @@
+/**
+# Copyright 2024 NVIDIA CORPORATION
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+**/
+
+package containerd
+
+import (
+	"os"
+	"path/filepath"
+	"sync"
+)
+
+type cgroupDriver string
+
+const (
+	cgroupDriverCgroupFS = cgroupDriver("cgroupfs")
+	cgroupDriverSystemd  = cgroupDriver("systemd")
+)
+
+var (
+	runningSystemd bool
+	detectSystemd  sync.Once
+)
+
+// getCgroupDriver returns the cgroup driver to use.
+// This is adapted from the containerd implementation.
+// https://github.com/containerd/containerd/blob/1fb1882c7de7239711c343c65fa266a2f1eba9db/internal/cri/server/runtime_config_linux.go#L63-L68
+func getCgroupDriver(root string) cgroupDriver {
+	detectSystemd.Do(func() {
+		fi, err := os.Lstat(filepath.Join(root, "/run/systemd/system"))
+		if err != nil {
+			return
+		}
+		runningSystemd = fi.IsDir()
+	})
+
+	if !runningSystemd {
+		return cgroupDriverCgroupFS
+	}
+	return cgroupDriverSystemd
```

Comment:
Just because sytemd is running doesn't mean that the systemd cgroup driver is in use by containerd. In older versions of containerd `cgroupfs` was the default and would be used even if systemd was in use by the machine in general. Are we OK, saying that we only support newer versions of containerd where the default cgroup driver is implied by the presence of systemd?

Reply:
Given that we only consume this setting if no config entry exists from which to infer it, does it make sense to change this function to `hasSystemd` or something similar? We would then only use it to set `SystemdCgroup` if not already specified.

In terms of only supporting newer containerd versions, I'm not sure what the right approach there should be. Since 1.6 is an LTS release, we should at least support this. Maybe this speaks to @tariq1890's point that we should not implement the automatic detection here and consider the implications more closely.

Reply:
Yeah, that sounds fine to me.

Let's call it `hasSystemd`, and then do what I suggest below about *explicitly* setting `SytemdCgroup` to `true` or `false` based on the results of it.

It still implies the existence of a "newish" containerd, but I think I'm OK with this.

Where the new semantics are -- "use whatever setting is in the config file, and if there is no setting, decide how to set it based on the existence of systemd on the machine".

This *may* break for older containerds, but there is a simple workaround (i.e. add an explicit setting in the config file) and over time we will converge on this as the "right fix".

Reply:
I have moved the systemd-specifics to #578 to keep this PR more focussed.